### DISCUSSION TITLE
Add stubs for _async methods

### DISCUSSION THREAD

**Discussion by dgilman-hrp** [auth0/auth0-python#749] [Ideas]
### Checklist

- [X] I have looked into the [Readme](https://github.com/auth0/auth0-python#readme) and [Examples](https://github.com/auth0/auth0-python/blob/master/EXAMPLES.md), and have not found a suitable solution or answer.
- [X] I have looked into the [API documentation](https://auth0-python.readthedocs.io/en/latest/) and have not found a suitable solution or answer.
- [X] I have searched the [issues](https://github.com/auth0/auth0-python/issues) and have not found a suitable solution or answer.
- [X] I have searched the [Auth0 Community](https://community.auth0.com) forums and have not found a suitable solution or answer.
- [X] I agree to the terms within the [Auth0 Code of Conduct](https://github.com/auth0/open-source-template/blob/master/CODE-OF-CONDUCT.md).

### Describe the problem you'd like to have solved

The built-in asyncify is very useful, but the way it's implemented means that typing checking (such as via Pyre) and IDEs have no idea these exist.

### Describe the ideal solution

Add function stubs for every method that supports async with type hinting.

**Comment by Avasam** [comment]
https://github.com/auth0/auth0-python/pull/678#issuecomment-2798399228
> Here's one way you could solve this (without changing anything at runtime): Generate stubs for all classes you expect to "Asyncify" as was talked here. But with the following differences:
> 
> * Generate it in a subfolder that doesn't exist at runtime (ie: `auth0._asyncified.*`)
> * Rename all the generated classes to help differentiate them (something like `class UsersWrapped(Users):` for base class `Users`)
> * Decorate all generated classes with `@typing.type_check_only`
> * Generate a massive overload for `auth0.asyncify.asyncify` that maps original classes to stub-only asyncified/wrapped classes.
> * Optional: Generate said stubs on install/editable install/wheel creation. That way they don't pollute the source and history.

> Like this: [Avasam/typeshed#43 (files)](https://github.com/Avasam/typeshed/pull/43/files#diff-2e519b89160890743c6aaa1d6e57c21160a2db3148ce891a8e5aa250b6bc9b62)

**Comment by kishore7snehil** [comment]
I believe this issue can be closed now as we already have a workaround discussed on this thread.

**Comment by Dreamsorcerer** [comment]
Isn't that workaround to wait for Avasam to merge the typeshed PR and then use types-auth0-python instead? Still looks to me like something needs to be done in this repo in order to provide types directly and close #612.

**Comment by kishore7snehil** [comment]
@Dreamsorcerer There was a PR for the same  where it was asked to add the `.py` files in the repo itself. I suggested to use the [script](https://github.com/auth0/auth0-python/pull/678#issuecomment-2737933954) to generate the async files.

That's the same approach @Avasam (correct me if I am wrong ) is pushing forward but with some specific tweaks to the generation logic.

If he is willing to add a PR to this repo, I can review it and take it forward.

**Comment by Dreamsorcerer** [comment]
> If he is willing to add a PR to this repo, I can review it and take it forward.

Of course. My point was just that this issue isn't closed until that PR gets created and merged.

**Comment by Avasam** [comment]
If you don't mind having a script to run once in a while to "keep the stubs in sync". I can migrate the script I made for https://github.com/python/typeshed/pull/13826 to over here.
Can probably include some CI check to ensure that running it wouldn't produce any change (ie: not forgotten by the contributor).
I see you have a https://github.com/auth0/auth0-python/blob/master/.pre-commit-config.yaml , it could also be added there.

Then I think you'll have everything you need to declare this library as "first-party typed".

**Comment by kishore7snehil** [comment]
@Avasam  Please feel free to open a PR. I can review it and will update you acordingly.

**Comment by kishore7snehil** [comment]
Closing this as resolved. The SDK `5.x `(generated via Fern) no longer uses the asyncify pattern that caused this issue. Async support is now first-class with explicit Async* client classes (e.g., AsyncUsersClient, AsyncRolesClient) that have fully typed async def methods. The package also ships with` py.typed` markers, so IDEs and type checkers work out of the box.

**Comment by StephenWithPH** [comment]
I believe https://github.com/python/typeshed/pull/14697 will have to land, followed by [6 months of waiting](https://github.com/python/typeshed/blob/main/CONTRIBUTING.md#third-party-library-removal-policy) for the out-of-date Typeshed stubs to cease masking the official typing.

**Reply by Dreamsorcerer** [reply]
The typeshed stubs are installed through a package, so just uninstall the package.

**Reply by Avasam** [reply]
@Dreamsorcerer
> The typeshed stubs are installed through a package, so just uninstall the package.

Some IDEs and type-checkers will bundle them.

pyright does
pylance (VSCode Python language server built on top of pyright) does https://github.com/microsoft/python-type-stubs on top of that (temporarily uninstallable by deleting the files until next updated)
From what I understand, PyCharm does something similar.

mypy won't (it'll recommend known stub install if the library is not typed)
idk about ty, pyrefly and zuban
